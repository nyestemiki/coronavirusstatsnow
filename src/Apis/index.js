// Apis
export { default as getData } from './getData';
export { default as getCountries } from './getCountries';
