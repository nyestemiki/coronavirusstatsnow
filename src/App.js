import React, { Component } from 'react';
import { Style, Top, Select, Language, Data, DataContainer, Number, Title, Updated } from './Style';
import localizationData from './localization';

const formatNumber = num => 
  num.toString().replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1,');

export default class App extends Component {
  state = {
    data: {},
    countries: {},
    date: "",
    lang: "en"
  }

  setLanguage = event => {
    this.setState({ lang: event.target.dataset.lang });
  }

  localization = word => localizationData[word] && localizationData[word][this.state.lang];

  updateDate = () => {
    const now = new Date();
    const mdy = now.toLocaleDateString();
    const hour = now.getHours();
    const minute = now.getMinutes();

    const date = mdy + " " + hour + ":" + (minute > 9 ? minute : "0" + minute);

    this.setState({ date });
  }

  selectCountry = () => {
    const selector = document.querySelector('#countySelector');
    const countryCode = selector.options[selector.selectedIndex].value;

    if (countryCode === 'world') {
      fetch('https://covid19.mathdro.id/api')
        .then(res => res.json())
        .then(data => this.setState({ data }));
    } else {
      fetch(`https://covid19.mathdro.id/api/countries/${countryCode}`)
        .then(res => res.json())
        .then(data => this.setState({ data }));
    }

    this.updateDate();
  }

  componentDidMount() {
    fetch('https://covid19.mathdro.id/api')
      .then(res => res.json())
      .then(data => this.setState({ data }));
    
    fetch('https://covid19.mathdro.id/api/countries')
      .then(res => res.json())
      .then(countries => this.setState({ countries: countries.countries }));

    this.updateDate();
  }

  render() {
    if (this.state.data.length === 0 || this.state.countries.length === 0) {
      return <p>Loading...</p>
    }

    return (
      <Style>
        <Top>
          <Select onChange={this.selectCountry} id="countySelector">
            <option key="world" value="world">{this.localization("World") || "World"}</option>
            {Array.from(this.state.countries).map(({ name, iso2 }) => 
              <option value={iso2} key={iso2}>{name}</option>
            )}
          </Select>
          <Language>
            <img data-lang="en" alt="en" src="/imgs/en.jpg" onClick={this.setLanguage}/>
            <img data-lang="hu" alt="hu" src="/imgs/hu.jpg" onClick={this.setLanguage}/>
            <img data-lang="ro" alt="ro" src="/imgs/ro.jpg" onClick={this.setLanguage}/>
            <img data-lang="de" alt="de" src="/imgs/de.jpg" onClick={this.setLanguage}/>
          </Language>
        </Top>

        <Data>
          <DataContainer>
            <Number>
            {this.state.data.confirmed && 
              formatNumber(this.state.data.confirmed.value)}
            </Number>
            <Title>{this.localization("Confirmed") || "Confirmed"}</Title>
          </DataContainer>
          <DataContainer>
            <Number>
              {this.state.data.recovered && 
                formatNumber(this.state.data.recovered.value)}
            </Number>
            <Title>{this.localization("Recovered") || "Recovered"}</Title>
          </DataContainer>
          <DataContainer>
            <Number>
              {this.state.data.deaths && formatNumber(this.state.data.deaths.value)}
            </Number>
            <Title>{this.localization("Deaths") || "Deaths"}</Title>
          </DataContainer>
        </Data>
        <Updated>
          <span>{this.localization("Updated") || "Updated"}</span>
          <span>{this.state.date}</span>
        </Updated>
      </Style>
    );
  }
}