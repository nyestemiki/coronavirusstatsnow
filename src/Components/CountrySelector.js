import React, { useContext} from 'react';
import LanguageContext from '../Context/LanguageContext';
import { Select } from '../Style';
import { localization } from '../Utils/localization';

export default function({ countries, selectCountry }) {
    const language = useContext(LanguageContext);

    return (
        <Select id="countySelector" onChange={event => 
            selectCountry(event.target[event.target.selectedIndex].value)
        }>
            <option key="world" value="world">
                {localization("World", language) || "World"}
            </option>

            {Array
                .from(countries)
                .map(({ name, iso2 }) => 
                    iso2 && <option value={iso2} key={iso2}>{name}</option>
            )}
        </Select>
    );
}