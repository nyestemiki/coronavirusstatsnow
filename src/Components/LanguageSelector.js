import React from 'react';
import { Languages } from '../Style';

export default function({ selectLanguage }) {
    const languages = ["en", "hu", "ro", "de"]

    return (
        <Languages>
            {languages.map(language => (
                <img 
                    data-lang={language} 
                    key={language}
                    alt={language} 
                    src={`/imgs/${language}.jpg`}  
                    onClick={() => selectLanguage(language)}/>
            ))}
        </Languages>
    );
}