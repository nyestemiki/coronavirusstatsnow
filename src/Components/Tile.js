import React from 'react';
import localization from '../Utils/localization';
import { formatNumber } from '../Utils/format';
import { DataContainer, Number, Title } from '../Style';

// Tile
export default ({ title, number }) => (
    <DataContainer>
        <Number>{formatNumber(number)}</Number>
        <Title>{localization(title)}</Title>
    </DataContainer>
)
