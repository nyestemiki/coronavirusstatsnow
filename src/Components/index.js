export { default as CountrySelector } from './CountrySelector';
export { default as LanguageSelector } from './LanguageSelector';
export { default as Tile } from './Tile';
export { default as Tiles } from './Tiles';
export { default as Updated } from './Updated';
