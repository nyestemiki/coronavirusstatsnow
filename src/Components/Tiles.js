import React from 'react';
import { Data } from '../Style';
import Tile from './Tile';

export default function({ language, data }) {
    return (
        <Data>
          <Tile 
            title="Confirmed" 
            language={language}
            number={data.confirmed?.value}/>

          <Tile 
            title="Recovered" 
            language={language}
            number={data.recovered?.value}/>
            
          <Tile 
            title="Deaths" 
            language={language}
            number={data.deaths?.value}/>
        </Data>
    )
}