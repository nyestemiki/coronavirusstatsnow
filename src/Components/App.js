import React, { useState, useEffect } from 'react';
import { Style, Top } from '../Style';
import { formatDate } from '../Utils/format';
import CountrySelector from './CountrySelector';
import LanguageSelector from './LanguageSelector';
import Tiles from './Tiles';
import Updated from './Updated';
import LanguageContext from '../Context/LanguageContext';
import getData from '../Apis/getData';
import getCountries from '../Apis/getCountries';

export default function() {
  const [data, setData] = useState({});
  const [countries, setCountries] = useState({});
  const [date, setDate] = useState("");
  const [language, setLanguage] = useState("en");

  async function selectCountry(countryCode) {
    let newData = null;
    
    if (countryCode === 'world') {
      newData = await getData();
    } else {
      newData = await getData(countryCode);
    }
    
    setData(newData);

    setDate( formatDate(new Date()) );
  }

  useEffect(() => {
    async function loadAllData() {
      setData(await getData());

      setCountries(await getCountries());
    }

    loadAllData();

    setDate( formatDate(new Date()) );
  }, []);

  if (data.length === 0 || countries.length === 0) {
    return <p>Loading...</p>
  }

  return (
    <LanguageContext.Provider value={language}>
      <Style>
        <Top>
          <CountrySelector selectCountry={selectCountry} countries={countries}/>
          <LanguageSelector selectLanguage={setLanguage}/>
        </Top>
        <Tiles data={data}/>
        <Updated date={date}/>
      </Style>
    </LanguageContext.Provider>
  );
}
