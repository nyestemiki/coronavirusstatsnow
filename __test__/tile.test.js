import { Tile } from '../src/Components';
import { shallow } from 'enzyme';

const fakeTile = { 
    title: "title", 
    number: 123456789 
};

describe('<Tile/>', () => {
    it('displays title properly', () => {
        const wrapper = shallow(<Tile {...fakeTile}/>);
        const Title = wrapper.find('Style__Title');

        expect(Title.dive().text()).toBe(fakeTile.title);
   }); 

   it('renders countup properly', () => {
        const wrapper = shallow(<Tile {...fakeTile}/>);
        const CountUp = wrapper.find('CountUp');

        expect(CountUp).toBeTruthy();
    }); 
});
