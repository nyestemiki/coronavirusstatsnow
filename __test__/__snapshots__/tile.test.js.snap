// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tile/> matches snapshot 1`] = `
<Style__DataContainer>
  <Style__Number>
    <CountUp
      decimal="."
      decimals={0}
      delay={0}
      duration={1}
      easingFn={null}
      end={123456789}
      formattingFn={null}
      onEnd={[Function]}
      onPauseResume={[Function]}
      onReset={[Function]}
      onStart={[Function]}
      onUpdate={[Function]}
      prefix=""
      preserveValue={false}
      redraw={false}
      separator=","
      start={0}
      startOnMount={true}
      suffix=""
      useEasing={true}
    >
      <Component />
    </CountUp>
  </Style__Number>
  <Style__Title>
    title
  </Style__Title>
</Style__DataContainer>
`;
